taichi/ui/common/field_info.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
+#include "taichi/ui/utils/utils.h"
+
+#include "taichi/ir/type_utils.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+enum class FieldSource : int {
+  TaichiCuda = 0,
+  TaichiX64 = 1
+  // support np array / torch tensor in the future?
```

Comment:
Definitely should support fields from vulkan, and maybe OpenGL

Reply:
Well yes. But right now there's no API to expose the corresponding buffer for those backends. For LLVM backends，we have ti.get_addr. I think we should start working on a more systematic way of interop-ing once we have SNode Descriptors ready.